futures-util/src/future/shared.rs

```diff
@@ -19,75 +19,101 @@
 //! # }
 //! ```
 
-use std::{error, fmt, mem, ops};
+use slab::Slab;
+
+use std::fmt;
 use std::cell::UnsafeCell;
+use std::marker::Unpin;
+use std::mem::PinMut;
 use std::sync::{Arc, Mutex};
 use std::sync::atomic::AtomicUsize;
 use std::sync::atomic::Ordering::SeqCst;
-use std::collections::HashMap;
+use std::task::local_waker_from_nonlocal;
 
-use futures_core::{Future, Poll, Async};
-use futures_core::task::{self, Wake, Waker, LocalMap};
+use futures_core::{Future, Poll};
+use futures_core::task::{self, Wake, Waker};
 
 /// A future that is cloneable and can be polled in multiple threads.
 /// Use `Future::shared()` method to convert any future into a `Shared` future.
 #[must_use = "futures do nothing unless polled"]
 pub struct Shared<F: Future> {
     inner: Arc<Inner<F>>,
-    waiter: usize,
+    waker_key: usize,
+}
+
+struct Inner<F: Future> {
+    future_or_output: UnsafeCell<FutureOrOutput<F>>,
+    notifier: Arc<Notifier>,
 }
 
-impl<F> fmt::Debug for Shared<F>
-    where F: Future + fmt::Debug,
-          F::Item: fmt::Debug,
-          F::Error: fmt::Debug,
-{
+struct Notifier {
+    state: AtomicUsize,
+    wakers: Mutex<Slab<Option<Waker>>>,
```

Comment:
I suggest putting the Slab into an option. That way it can be dropped immediately and efficiently after the future completes.

`wakers: Mutex<Option<Slab<Option<Waker>>>>`

Reply:
I don't need an option to free it-- I can use `mem::replace` with a new slab.

Reply:
I don't think that will work. `Slab::remove` will panic if you call it with a key that does not exist.

Reply:
I mean the call to `Slab::remove` in `drop`